### DISCUSSION TITLE
How to build create Action to build legacy .NET Framework 4 project?

### DISCUSSION THREAD

**Discussion by pf-uvonceumer** [community/community#25253] [Actions]
<p>Hi,</p>
<p>I am struggling to create an action that builds my .NET Framework 4 solution.</p>
<p>Here is what I have:</p>
<pre><code class="lang-auto">name: CI

on:
  workflow_dispatch:    
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

jobs:
  build:

    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
    
    - name: Setup MSBuild
      uses: microsoft/setup-msbuild@v1

    - name: Setup NuGet
      uses: NuGet/setup-nuget@v1.0.5
      
    - name: setup-msbuild
      uses: microsoft/setup-msbuild@v1.1

    - name: Restore Packages
      run: nuget restore MySolution.sln

    - name: Build solution
      run: msbuild MySolution.sln -t:rebuild -property:Configuration=Release
</code></pre>
<p>And here is the error message I am getting:</p>
<pre><code class="lang-auto">
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(1220,5): error MSB3644: The reference assemblies for .NETFramework,Version=v4.0 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks [D:\a\MySolution-Legacy\MySolution-Legacy\MySolution.Tests\MySolution.Tests.csproj]

[21](https://github.com/Pepperl-Fuchs-SE-PIT/ShopfloorDataCache-Legacy/runs/5409613295?check_suite_focus=true#step:4:21)Done Building Project "D:\a\MySolution-Legacy\MySolution-Legacy\MySolution.Tests\MySolution.Tests.csproj" (Rebuild target(s)) -- FAILED.

[22](https://github.com/Pepperl-Fuchs-SE-PIT/ShopfloorDataCache-Legacy/runs/5409613295?check_suite_focus=true#step:4:22)Done Building Project "D:\a\MySolution-Legacy\MySolution-Legacy\MySolution.sln" (rebuild target(s)) -- FAILED.
</code></pre>
<p>So it seems like .NET v4.0 is missing on the system. However, I have no idea on how to setup the runner to support .NET 4.0.</p>
<p>Any help would be greatly appreciated.</p>
<p>Cheers,<br>
Ulf</p>

**Comment by mickeygousset** [comment]
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/actions/setup-dotnet">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/actions/setup-dotnet" target="_blank" rel="noopener">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/20031479/181089883-e62f2952-3f88-4cfc-897f-004e4fe78136.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/6/36d1880da28d21c705c527170b000af222f28324_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/3/6/36d1880da28d21c705c527170b000af222f28324_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/6/36d1880da28d21c705c527170b000af222f28324.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/6/36d1880da28d21c705c527170b000af222f28324_2_10x10.png"></div>

<h3><a href="https://github.com/actions/setup-dotnet" target="_blank" rel="noopener">GitHub - actions/setup-dotnet: Set up your GitHub Actions workflow with a...</a></h3>

  <p>Set up your GitHub Actions workflow with a specific version of the .NET core sdk - GitHub - actions/setup-dotnet: Set up your GitHub Actions workflow with a specific version of the .NET core sdk</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Try adding this action to your workflow</p>

**Comment by pf-uvonceumer** [comment]
<p>Already tried this. This only seems to work for .NET CORE / SDK style projects. Not for legacy .NET Framework projects.</p>

**Comment by pf-uvonceumer** [comment, marked as answer]
<p>So, I found out what was wrong.</p>
<p>I need to use older <code>windows-2019</code> runners. It seems like the <code>windows-2022</code> runners do not have .NET 4.0 SDK installed.</p>

**Reply by kinokopio** [reply]
How should the code be written?

**Reply by pf-uvonceumer** [reply]
like this:

`runs-on: windows-2019`

**Reply by jitendrakumarsuthar** [reply]
100% worked for me, thanks a lot @pf-uvonceumer.
For SonarQube GitHub Action integration, inside the .yml file replace it at `runs-on: windows-latest` with `runs-on: windows-2019`

**Comment by ljnath** [comment]
I have a .NET class project which targets multiple versions of .NET framework, .NET Core (inc. 8.0) and .NET standard.
For latest version of .NET Core I need to use `windows-2022`; however .NET Framework 4.0 is not supported in `windows-2022` and only supported in `windows-2019`.

Can anyone please suggest some way to successful run my GitLab action file. My current action file is here https://github.com/ljnath/RandomString4Net/blob/master/.github/workflows/dotnet-CI-workflow.yml

**Reply by Xriuk** [reply]
I also have a multi-version .NET library, since apparently it is not possible to install additional .NET framework my solution is to use a `windows-2019` agent and install additional .NET/.NET Core:
https://github.com/Xriuk/NeatMapper/blob/main/.github/workflows/nuget.yml
